pkg/controllers/gateway/gateway_controller.go

```diff
@@ -70,88 +88,237 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
-		applyClusterSyncerAnnotations(gateway, clusters)
-		statusConditions = append(statusConditions, metav1.Condition{
-			LastTransitionTime: metav1.Now(),
-			Message:            fmt.Sprintf("Gateways configured in data plane clusters - [%v]", strings.Join(clusters, ",")),
-			Reason:             string(gatewayv1beta1.GatewayConditionProgrammed),
-			Status:             metav1.ConditionTrue,
-			Type:               string(gatewayv1beta1.GatewayConditionProgrammed),
-			ObservedGeneration: previous.Generation,
-		})
-	}
-
-	if !reflect.DeepEqual(gateway.Annotations, previous.Annotations) {
-		log.Info("Updating Gateway annotations", "gateway", gateway)
+	// Update gateway spec/metadata
+	if !reflect.DeepEqual(gateway, previous) {
+		log.Info("Updating Gateway Spec", "gatewaySpec", gateway.Spec, "previousSpec", previous.Spec)
+		log.Info("Updating Gateway ObjectMeta", "gatewayObjectMeta", gateway.ObjectMeta, "previousObjectMeta", previous.ObjectMeta)
 		err = r.Update(ctx, gateway)
 		if err != nil {
-			return ctrl.Result{}, err
+			log.Error(err, "Error updating Gateway")
 		}
 	}
 
-	gateway.Status.Conditions = statusConditions
+	// Update status
+	gateway.Status.Conditions = buildStatusConditions(gateway.Status, previous.Generation, clusters, acceptedStatus, programmedStatus)
 	if !reflect.DeepEqual(gateway.Status, previous.Status) {
-		log.Info("Updating Gateway status", "gatewayStatus", gateway.Status)
+		log.Info("Updating Gateway status", "gatewaystatus", gateway.Status, "previousstatus", previous.Status)
 		err = r.Status().Update(ctx, gateway)
 		if err != nil {
-			return ctrl.Result{}, err
+			log.Error(err, "Error updating Gateway status")
 		}
 	}
 
-	return ctrl.Result{}, nil
+	if requeue {
+		return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, reconcileErr
+	}
+	return ctrl.Result{}, reconcileErr
 }
 
-func applyClusterSyncerAnnotations(gateway *gatewayv1beta1.Gateway, clusters []string) {
-	for _, cluster := range clusters {
-		gateway.Annotations[fmt.Sprintf("%s/%s", ClusterSyncerAnnotation, cluster)] = "True"
+// Configures Gateway tls & dns for each cluster it targets.
+// Returns the programmed status, a list of clusters that were programmed, if the gateway should be requeued, and any error
+func (r *GatewayReconciler) reconcileGateway(ctx context.Context, previous gatewayv1beta1.Gateway, gateway *gatewayv1beta1.Gateway) (metav1.ConditionStatus, []string, bool, error) {
+	log := log.FromContext(ctx)
+
+	clusters := selectClusters(*gateway)
+	// Don't do anything else until at least 1 cluster matches.
+	if len(clusters) == 0 {
+		// TODO: Handle any cleanup if there were previously clusters
+		return metav1.ConditionFalse, clusters, false, nil
 	}
+	trafficAccessor := traffic.NewGateway(gateway)
+	gatewayHelper := trafficAccessor.(GatewayHelper)
+	allHosts := trafficAccessor.GetHosts()
+	hostsWithMZ := []string{}
+
+	// First, filter out hosts that are *not* in a Managed Zone
+	// Those can be left alone
```

Comment:
It looks like this comment might be out of date? I think this for block is now generating an empty DNS record for each subdomain that exists in a valid managed zone?

Reply:
oh I see it is also building hostsWithMZ at the end.

Reply:
Yes, exactly